docs/lang/articles/misc/ndarray_android.md

```diff
@@ -0,0 +1,449 @@
+              2
+            ],
+            "shape_offset_in_bytes_in_args_buf": 96
+          },
+          "2": {
+            "field_dim": 1,
+            "is_scalar": false,
+            "element_shape": [
+              2
+            ],
+            "shape_offset_in_bytes_in_args_buf": 128
+          }
+        },
+        "used.arr_arg_to_bind_idx": {
+          "0": 5,
+          "1": 6,
+          "2": 4
+        }
+      }
+    },
+    "kernel_tmpls": {},
+    "fields": [],
+    "root_buffer_size": 0
+  }
+}
+```
+The following provides a detailed description of the keys in the generated JSON file:
+
+`aot_data`: The overarching JSON object.
+ - `kernels`: All Taichi kernels.
+	- `$(kernel_name)`: Name of a specific Taichi kernel.
+		- `tasks`: A JSON array of the generated compute shaders.
+			- `name`: Name of a specific compute shader.
+			- `src`: Relative path to the shader file.
+			- `workgroup_size`: N/A
+			- `num_groups`: N/A
+		- `arg_count`: Number of the arguments that the Taichi kernel takes.
+		- `ret_count`: Number of the values that the Taichi kernel returns.
+		- `args_buf_size`: The size of `args_buf` in Bytes.
+		- `ret_buf_size`: The size of `ret_buf` in Bytes.
+		- `scalar_args`: Scalar arguments that the kernel takes.
+		- `arr_args`: Shape information of the Ndarrays in the kernel.
+			- `$(arg_index)`: Argument index of an Ndarray
+				- `field_dim`: The dimension of the Ndarray.
+				- `is_scalar`: Whether the elements in the Ndarray are scalar.
+				- `element_shape`: An `int` array indicating the shape of each element in the Ndarray.
+				- `shape_offset_in_bytes_in_args_buf`: The offset of the Ndarray's shape information in `args_buf`.
+		- `used.arr_arg_to_bind_idx`: A map specifying the SSBO to bind for a given Ndarray. For example, `"1": 5` (line 48) binds the `vel` Ndarray to the binding index `5`.
+
+*Well, we hope you were not overwhelmed with that much information coming in all at once. In the following section, we will revisit the JSON file, as well as provide tables and graphs that help illustrate some of the concepts and notions listed above.*
+
+### Prepare SSBO and shape information
+
+Before executing the GLES compute shaders in your runtime program, you need to get all your resources ready, including:
+
+- Bind SSBO for the corresponding buffer
+- Bind SSBO for each Ndarray
+- Fill `args_buffer` with Ndarray shape information
+
+#### Bind SSBO for the corresponding buffer
+
+The following table lists the buffers commonly used in a Taichi program together with their binding indexes:
+
+| **Buffer**    | **Global/kernel-spedific** | **Storing**                                                  | **Binding index** |
+| ------------- | -------------------------- | ------------------------------------------------------------ | ----------------- |
+| `root_buffer` | Global                     | All fields with fixed offsets and of fixed sizes.            | `0`               |
```

Comment:
You only use root_buffer when your program has mix using fields and Ndarrays (or only has fields).

Reply:
You *only* need to bind an SSBO for `root_buffer` if your Taichi script uses at least one field. Skip this step if your script does not involve field.